test/e2e/far_e2e_test.go

```diff
@@ -0,0 +1,208 @@
+			if cond, errBoot = getKubeletReadyCondition(testNode.Name); errBoot != nil {
+				log.Error(errBoot, "Can't get boot time of the node")
+			}
+			nodeBootTimeBefore = cond.LastTransitionTime
+			far = createFAR(testNode.Name, fenceAgentIPMI, testShareParam, testNodeParam)
+		})
+
+		AfterEach(func() {
+			deleteFAR(far)
+		})
+
+		When("running FAR to reboot node ", func() {
+			It("should execute the fence agent cli command", func() {
+				By("checking the CR has been created")
+				farCR := &v1alpha1.FenceAgentsRemediation{}
+				ExpectWithOffset(offsetExpect, k8sClient.Get(context.Background(),
+					client.ObjectKey{Name: testNode.Name, Namespace: testNamespace}, farCR)).ToNot(HaveOccurred())
+
+				By("checking the command has been executed successfully")
+				checkFarLogs(cli.SuccessCommandLog)
+
+				By("checking the node's boot time after running the FA")
+				if !nodeBootTimeBefore.IsZero() {
+					wasNodeRebooted(testNode.Name, nodeBootTimeBefore)
+				} else {
+					Skip("we couldn't get the boot time of the node prior to FAR CR, thus we can't verify if it was rebooted")
+				}
+			})
+		})
+	})
+})
+
+// createFAR assign the input to FenceAgentsRemediation object, create CR it with offset, and return the CR object
+func createFAR(nodeName string, agent string, sharedparameters map[v1alpha1.ParameterName]string, nodeparameters map[v1alpha1.ParameterName]map[v1alpha1.NodeName]string) *v1alpha1.FenceAgentsRemediation {
+	far := &v1alpha1.FenceAgentsRemediation{
+		ObjectMeta: metav1.ObjectMeta{Name: nodeName, Namespace: testNamespace},
+		Spec: v1alpha1.FenceAgentsRemediationSpec{
+			Agent:            agent,
+			SharedParameters: sharedparameters,
+			NodeParameters:   nodeparameters,
+		},
+	}
+	ExpectWithOffset(offsetExpect, k8sClient.Create(context.Background(), far)).ToNot(HaveOccurred())
+	return far
+}
+
+// deleteFAR delete the CR with offset
+func deleteFAR(far *v1alpha1.FenceAgentsRemediation) {
+	EventuallyWithOffset(offsetExpect, func() error {
+		err := k8sClient.Delete(context.Background(), far)
+		if apiErrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}, timeout, pollInterval).ShouldNot(HaveOccurred(), "failed to delete far")
+}
+
+// getKubeletReadyCondition return the  condition of Node kubelet, otherwise return an error
+func getKubeletReadyCondition(nodeName string) (corev1.NodeCondition, error) {
+	emptyCondition := corev1.NodeCondition{}
+	node, err := clientSet.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return emptyCondition, err
+	}
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == corev1.NodeReady {
+			return condition, nil
+		}
+	}
+	return emptyCondition, fmt.Errorf("Node %s is not ready", nodeName)
+}
+
+// checkFarLogs get the FAR pod and check whether its logs has logString
+func checkFarLogs(logString string) {
+	By("checking logs")
+	var pod *corev1.Pod
```

Comment:
There is already a `By` section before calling this function, maybe we can get rid of this one.